Add sunos support

nshalman opened the pull request:
### Wireguard (wireguard-go) for sunos (illumos/Solaris)

- [x] reliability testing via Tailscale (Tailscale for both Solaris and illumos can be built on top of this work and it has been quite reliable.)
- [x] a wg-quick(8) implementation, like we have for FreeBSD and OpenBSD and such: WireGuard/wireguard-tools#17

### Original Description

A currently working cleanup of the code by @jclulow from https://github.com/jclulow/wireguard-go-illumos-wip squashed down to a single commit.

This code works with the caveats noted in the commit message, and with additional modifications, can be added to the https://github.comtailscale/wireguard-go tree and be used to build mostly working tailscale binaries.

I barely know what I'm doing here, so any help carrying this across the finish line would be greatly appreciated.

nshalman commented:
Pinging @crawshaw and @danderson Per https://github.com/tailscale/tailscale/issues/697#issuecomment-677823895

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+ * WIREGUARD/TUN INTERFACE
+ */
+
+type NativeTun struct {
```
It would be good Go style to add docstrings for exported names, but the rest of this package doesn't do that either. If you're feeling zealous, adding them would make this codebase more idiomatic Go. e.g.

`// NativeTun is the OS-specific implementation of a tun interface.`

Not required though, since as I said the rest of this package doesn't follow that convention.

zx2c4 replied:
Please don't do that here, but if you (@danderson) want to go through and add documentation to places in a consistent manner, please submit that as a different PR; it'd be welcome.

danderson replied:
I meant adding docstrings just to the illumos files, rather than bundle in unrelated changes to the other files. Happy to ship documentation for the other bits separately, yeah.

zx2c4 replied:
Would still love to have mentioned docs at some point.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+	errors chan error
+}
+
+// type Device interface {
```
AFAICT, this is here to try and hint that the following methods are implementations of this interface, to steer readers in the right direction. The common way to do this in Go code is to have docstrings on the methods that just say `// MethodName implements the Device interface.` (assuming the methods are trivial enough that they don't warrant a full docstring detailing their function and semantics - as is mostly the case here)

Agreed that I would remove this comment block, and instead either put docstrings on individual methods, or follow the established style of the rest of the package and let the reader connect the dots to the Device interface.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+		return 0, err
+
+	default:
+		return tun.read_tun(buf[offset:])
```
Yup, Go uses camelCase (or CamelCase for exported names). Unless you're explicitly calling out to a C function, you should use that style. It's a common gotcha for OS C programmers coming to Go :)

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+}
+
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	return tun.write_tun(buf[offset:])
```
There's discussion of that further down, but if these helpers need to continue doing deep unsafe things, the encapsulation of that into a helper is nice in that it isolates the unsafety a little. Hopefully we can dispense with the unsafe stuff instead.

If we do end up keeping the helpers, I would put them right next to their parent functions, so that a reader can go straight from the friendly Go Read/Write's implementation into the tricky internals, if they're trying to follow the entirety of the read/write path.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
+		ip_muxid, err := get_ip_muxid(tun.ip_fd, tun.name)
```
Same thing with the variable names, should be ipMuxID. In general, `snake_case` shouldn't be used anywhere in Go except when calling out to C code.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+	 * descriptor number as the fd number is passed (with no other
+	 * information) through the environment to the child.  For now, use the
+	 * "-f" flag.
+	 */
```
Did some quick rooting around in illumos-gate, but couldn't find net/if_tun.h (which seems to be the expected include file for both Solaris and Illumos). I also couldn't find it in SmartOS or OmniOS's forks, though I did find https://github.com/joyent/smartos-live/issues/330 .

This might be a minor issue for adding tuntap things to x/sys/unix, in that you'll have to manually define the constant blocks and whatnot, rather than being able to autogenerate them from the C headers.

Presumably that missing include also has the answer to "how do I get the interface name for this fd".

zx2c4 replied:
Any updates on these investigations?

nshalman replied:
Per the thread below on "cheating", I have thus far been unable to find an IOCTL for getting back the PPA.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
```
I presume this is because there's no way of requesting a specific interface name, so you hardcoded this backstop in? If we can track down if_tun.h (or the corresponding kernel side implementation), I expect it'll have an ioctl for setting names as well, at which point we can remove this.

zx2c4 replied:
I'd be surprised if at least tunX wasn't allowed. Can you dig around and become certain one way or another?

zx2c4 replied:
Any update here?

nshalman replied:
tunX is allowed and now implemented.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+	}
+
+	name = fmt.Sprintf("tun%d", ppa)
+	fmt.Printf("device %v\n", name)
```
Leftover debug printf, remove?

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+
+	/*
+	 * Open another temporary file descriptor to the TUN driver.
+	 * XXX It's not clear if this is actually needed, or if we could
```
Unless there's surprising semantics in the Illumos API surface, seems like it should be fine to reuse the device FD you extracted from the os.File.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+func push_ip(fd int) (error) {
+	/*
+	 * We need a C string with the value "ip".
+	 * XXX Is a Golang "string" implicitly NUL-terminated?
```
It's okay. The usual way to do this is to have several `IoctlSet<go-type>` functions in x/sys/unix, and have x/sys/unix do the Go-to-C string management (usually via cgo's `C.CString` helper), struct packing, and so forth. That way the callsite here is just `unix.IoctlSetString(fd, I_PUSH, "ip")` (at which point the function is simple enough that you can inline push_ip at its callsite)

danderson replied:
Same thoughts on the other ioctl calls below, I'm not going to call out each one individually. I think each should become an `IoctlSet<type>` function in x/sys/unix's Illumos implementation, at which point a lot of this glue evaporates and can likely be inlined in CreateTUN.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+}
+
+func get_ip_muxid(fd int, name string) (int, error) {
+	// struct lifreq { /* 0x178 bytes */
```
Moving this into x/sys/unix would mostly take care of that. If this struct is defined in a .h somewhere in Illumos, x/sys/unix's generator machinery will construct the matching Go struct for you, including padding and all that good stuff. See also comment below on how to plumb ioctl calls into x/sys/unix and avoid having all this glue in wireguard-go itself.

nshalman replied:
See golang/sys#100

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+
+	var flags int32 = 0
+
+	_, err := getmsg(int(tun.tunFile.Fd()), uintptr(0),
```
Do STREAMS offer any kind of batched read/write API? Or does getmsg/putmsg handle batched delivery under the hood? I ask because "one syscall per packet" ends up limiting tuntap performance on many platforms, and I'm wondering if illumos' use of STREAMS would let it sidestep that from the start.

nshalman replied:
@gco pointed me at `bufmod(4M)`
- https://docs.oracle.com/cd/E88353_01/html/E37851/bufmod-4m.html
- https://illumos.org/man/4M/bufmod

I haven't explored what implementing that would look like here.

zx2c4 replied:
Sounds like a nice optimization. Let's get things working right the vanilla way before we do things the wild way, though. It's already been a few years and this still hasn't shipped.

nshalman replied:
Agreed. I just wanted to leave the breadcrumb for that here rather than leaving it sitting in my Twitter DMs. :)

gco replied:
It will add latency, though.   I suspect that putting the STREAM in RMSGD mode and using readv with sufficiently large buffers will work to read multiple packets with one syscall.

zx2c4 commented:
> The other question we should resolve soon-ish is whether we're going to try for SMF integration in wg-quick, or go down the less-managed route that the BSDs took.

Thanks for offering to work on this. If you're big into Illumos, taking the lead on the wg-quick integration would be much appreciated. I'd prefer to keep the same semantics as all the other wg-quick implementations. If we want to later add deeper Illumos integration with SMF, that sounds fine, but would be a separate endeavor. For now, let's get everything all lined up.

Usually there are a few problems to solve in wg-quick porting, beyond obvious things:
- How do you change the DNS server in an exclusive way? Most implementations use r`esolvconf -m 0 -x`, which is preferable. Some use the dns hatchet (see contrib/), which is grotesque. Some just overwrite /etc/resolv.conf, which is insufficient. I assume Illumos has something nice and uniform we can use.
- How do we exempt the WireGuard socket from the default routing rule? Linux uses fwmarks, for example. [@nshalman - you'll want to add an implementation for this in mark_unix.go, by the way.]
- How do we overwrite the default route without removing the prior default route? Linux uses multiple routing tables. Some of the BSDs `&disown` a little route monitor in bash. Hopefully Illumos has a Linux-like solution that doesn't require background processes.
- How do you give an IP address to an interface without providing a "gateway" IP? Linux has the `dev wg0` notation. Some BSDs support a `-interface` switch to ifconfig. Others rely on gross hacks. Hopefully Illumos has something clean.

I would suggest reading all of the current wg-quick implementations, except perhaps the Android one, before starting, and then maybe start with one of the BSD implementations and tweak it.

If you have additional questions about `wg-quick` development, I suggest putting this discussion on the mailing list, where wg-quick development happens, and where mailing list lurkers periodically pop up with some nice domain specific knowledge.

zx2c4 commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,146 @@
+
+type UAPIListener struct {
+	listener net.Listener // unix socket listener
+	connNew  chan net.Conn
```
Oh, you were wondering about SMF to manage wireguard-go's lifetime, not wg-quick's. Gotcha. Same response here, though -- let's roll with the usual wireguard-go semantics and not try to do too much that's fancy right away.

danderson replied:
Okay, so that means just backgrounding wireguard-go, same as the other systems. Easy enough. @nshalman in that case, you should adapt the BSD code's use of kqueue to Illumos' file event ports. Unfortunately, that's more API surface that isn't exposed in x/sys/unix. We'd want to take inspiration from the PR at https://github.com/fsnotify/fsnotify/pull/263/files to get the right port primitives into x/sys/unix as well.

zx2c4 replied:
@nshalman - I too am wondering about where the file watcher stuff is for this. UAPI on other platforms is able to notify when the socket has been removed.

nshalman replied:
I've got open CRs to get the event ports APIs from the fsnotify PR into x/sys/unix and a fork of fsnotify that uses them and passes all integration tests. Would it make sense to use fsnotify or is there a preference to use event ports directly?

nshalman replied:
https://github.com/fsnotify/fsnotify/pull/371 is still a draft, but the underlying event ports support is upstream in x/sys/unix now.

nshalman replied:
Event Ports is in x/sys/unix and being used here.

zx2c4 commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,146 @@
+// +build illumos
```
Ack. So we'll keep it as a separate thing and monitor socket file deletion using an eventport. That's fine with me.

danderson commented:
> Thanks for offering to work on this. If you're big into Illumos, taking the lead on the wg-quick integration would be much appreciated. I'd prefer to keep the same semantics as all the other wg-quick implementations. If we want to later add deeper Illumos integration with SMF, that sounds fine, but would be a separate endeavor. For now, let's get everything all lined up.

Sounds good. In that case, the Illumos implementation of wg-quick is likely to be nearly identical to the OpenBSD implementation:

> * How do you change the DNS server in an exclusive way? Most implementations use r`esolvconf -m 0 -x`, which is preferable. Some use the dns hatchet (see contrib/), which is grotesque. Some just overwrite /etc/resolv.conf, which is insufficient. I assume Illumos has something nice and uniform we can use.

Illumos has no facilities for this that I could find (no resolvconf, and no Illumos-specific facility for managing DNS). It also lacks the piecemeal namespacing support that dns hatchet relies on, so we're stuck with replacing /etc/resolv.conf.

> * How do we exempt the WireGuard socket from the default routing rule? Linux uses fwmarks, for example. [@nshalman - you'll want to add an implementation for this in mark_unix.go, by the way.]

There's no policy routing facilities. Similar to OpenBSD, we'll have to add /32 or /128 routes to the endpoints.

> * How do we overwrite the default route without removing the prior default route? Linux uses multiple routing tables. Some of the BSDs `&disown` a little route monitor in bash. Hopefully Illumos has a Linux-like solution that doesn't require background processes.

Again, similar situation to the BSDs, no good solution. AFAICT, the current FreeBSD and OpenBSD scripts just add/delete routes per the wg-quick configuration and don't handle conflict at all. The route monitor updates endpoint routes to avoid loops, but does nothing to handle repairing the routing table on exit.

> * How do you give an IP address to an interface without providing a "gateway" IP? Linux has the `dev wg0` notation. Some BSDs support a `-interface` switch to ifconfig. Others rely on gross hacks. Hopefully Illumos has something clean.

Illumos's route(1M) supports -interface, so again should be the same as OpenBSD.

Based on this, once we've merged the initial Illumos support, I'll probably start with a copy of the OpenBSD implementation, then circulate that in the Illumos community to see if there's something we can do better.

nshalman commented:
Is my experimental change (https://github.com/golang/sys/compare/master...nshalman:illumos-wireguard) to x/sys/unix vaguely close to what it should be?
If so, I have an alternate branch that I will push into this one that clears out the ASM file and the corresponding code.
If not, additional guidance would definitely be welcome.
Thanks so much everyone!

nshalman commented on tun/asm_illumos_amd64.s:
```diff
@@ -0,0 +1,17 @@
+// System calls for amd64, Solaris are implemented in runtime/syscall_solaris.go
+//
+
+TEXT ·sysvicall6(SB),NOSPLIT,$0-88
```
That trick worked, but has been further obviated by golang/sys#100

nshalman commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,519 @@
+ * ILLUMOS SYSTEM CALL HELPERS
+ */
+
+//go:cgo_import_dynamic libc_ioctl ioctl "libc.so"
```
Help was received! golang/sys#100

nshalman commented:
With a lot of help from @danderson this is getting much closer to being ready to land.

danderson commented:
For the record - waiting until we get the upstream bits into x/sys/unix before I review this more closely again, so that we don't go through multiple rounds as review feedback alters the APIs.

teutat3s commented:
Just chiming in here to say I'm really looking forward to this! Thank you for hacking on `illumos support`!

If I can be of any help testing anything, I'd be more than happy to help.

nshalman commented:
This is currently stuck waiting for me to find time to clean up / refactor https://go-review.googlesource.com/c/sys/+/302831 to the point that it's ready for merging.  After that, it should be pretty straightforward to clean up this branch into something reasonable. No ETA unfortunately, but it's not abandoned either.

nshalman commented:
Sorry it's been so long. All necessary unsafe things have been moved to x/sys/unix. I imagine there's still more work to be done, but I think this is another moment where it would be a good time for more eyes.

zx2c4 commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+func UAPIListen(name string, file *os.File) (net.Listener, error) {
+
+	// wrap file in listener
+
```
Lots of whitespace, and comment probably isn't necessary.

zx2c4 commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,146 @@
+
+	uapi := &UAPIListener{
+		listener: listener,
+		connNew:  make(chan net.Conn, 1),
```
@danderson Feel free to submit a patch to the other platforms either before or after this patchset.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+// NativeTun is the OS-specific implementation of a tun interface.
+type NativeTun struct {
+	tunFile *os.File /* TUN device file */
+	ip_fd   int      /* IP device fd */
```
`ip_fd`  can probably be called something like `dgramFD`. But -- it's fairly rarely used, right? And only ever for specific control events? On the BSDs, we just open a new FD for every control request.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+// NativeTun is the OS-specific implementation of a tun interface.
+type NativeTun struct {
+	tunFile *os.File /* TUN device file */
```
These comments, `/* TUN device file */` are obvious enough that they probably aren't necessary. If you do think there's something non-obvious going on, use `//` comments for single-line struct member annotation.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	ip_fd   int      /* IP device fd */
+	name    string   /* Interface name */
+	mtu     int
+
```
No need for the newline.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	errors chan error
+}
+
+// Name implements the Device interface
```
This comment and all the others like it do not contribute anything meaningful. Just get rid of it.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+// Flush implements the Device interface
+func (tun *NativeTun) Flush() error {
+	// TODO: can flushing be implemented by buffering and using sendmmsg?
```
Maybe, but probably not worth the hassle.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	mtu     int
+
+	events chan Event
+	errors chan error
```
The errors channel is doing nothing in this implementation. Just get rid of it.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+// Read implements the Device interface
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
```
As mentioned above, `tun.errors` isn't doing anything. Just get rid of it.

You probably want a RWLock-protected `closed` boolean instead though.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	case err := <-tun.errors:
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(int(tun.tunFile.Fd()), nil, buf[offset:])
```
Would you be better off with caching the SysConn and using the safer closures it provides, instead of using Fd(), which is potentially racy?

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+// Write implements the Device interface
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	if err := unix.Putmsg(int(tun.tunFile.Fd()), nil, buf[offset:], 0); err != nil {
```
Same question about SyscallConn() instead of Fd().

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+// Close implements the Device interface
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
```
Just open a new DGRAM socket here instead.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+	if tun.tunFile != nil {
+		tun.tunFile.Close()
+		tun.tunFile = nil
```
If you're going to use `SyscallConn`, as suggested above, maybe don't set this to nil here.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	name    string   /* Interface name */
+	mtu     int
+
+	events chan Event
```
There's still no code to actually generate events -- MTU-change and interface up/down.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	 * To establish a "tun" interface, we need to open a few file
+	 * descriptors.
+	 */
+	ip_node := "/dev/udp"
```
`const udpNode = "/dev/udp"`

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+	 * descriptors.
+	 */
+	ip_node := "/dev/udp"
+	dev_node := "/dev/tun"
```
`const tunNode = "/dev/tun"`

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+		unix.Close(ip_fd)
+		return nil, fmt.Errorf("Could not open TUN (%s)", dev_node)
+	}
+	fd := int(tunFile.Fd())
```
Cache the SyscallConn() and use that, and avoid Fd(), perhaps, as discussed above.

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,271 @@
+
+	defer func() {
+		/*
+		 * XXX For now, we'll just send a link up event straight away.
```
Will need to be completed before merging.

nshalman replied:
What is supposed to go in here?

zx2c4 commented:
Thanks for your continued work on this. I did a pass at it, as requested, but it looks like lots of other comments from before are still unaddressed and there are still many `XXX:` and TODOs in the code. So this stuff should probably be finished up before this is merged.

nshalman reviewed (commented):
Replied to a few threads. I've had some promising results with Event Ports and fsnotify work.
Should be good building blocks for picking this back up and making forward progress.

nshalman commented:
Still some more cleanup to do, but we're getting very very close.
Not marking for review yet because there are still some unaddressed comments I should take care of.

nshalman commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,146 @@
+// +build illumos
```
I think go fmt might have opinions about this...

zx2c4 commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
```
Can you test this all and remove the XXX comments before I do a full review? If you have questions or need help with certain parts, feel free to say so in ordinary comments.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+		statusMTU int
+	)
+
+	tunIfindex, err := func() (int, error) {
```
I know you copied this from the freebsd implementation, but it doesn't make much sense to me to define an in-place closure and execute it once at the time of definition. More idiomatic would be to have this either as a separate method on NativeTun, so this callsite turns into `tun.index()`. Or, since it's only used once, remove the surrounding closure and make it straight-line code that does the InterfaceByName lookup followed by extracting the index. The resulting code would be shorter too, since you'd have a single error check instead of a pair.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+		return
+	}
+
+	defer close(tun.events)
```
Since the tunIfIndex logic can error out, you want to move this defer above it, so the channel closes regardless of which error path causes an exit.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+		if err != nil {
+			//tun.errors <- err
+			//return
+			// XXX returning this error breaks everything??
```
This suggests that the struct being parsed is not the same on illumos and freebsd, where you got this code from. That would result in finding nonsensical indices, and having the interface lookup fail despite the tunnel still existing. Looking at golang.org/x/net/route, which has parsers for these RIB messages, it looks like there's 2-3 variants that put the ifindex at different offsets.

Easy fix for this is to look up in the illumos source what offset the ifindex sits at, and adjust the unsafe cast above.

Longer term, it might be nice to teach x/net/route to parse illumos RIB messages, and switch wireguard-go to using those parsers for all OSes, to remote another use of unsafe from this codebase. I think that should be a separate refactoring PR that makes the change across all platforms though, for this particular PR I think it's better to stick to the existing parsing strategy, which all the other non-linux unices also use.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+
+		// illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint. Hand-cast things to help
+		// the compiler figure it out.
```
Get rid of the last sentence, it's a natural conclusion from the prior sentence and the cast. It's also not quite helping the compiler figure it out: in Go, casts between integer types are permitted, but not implicit. The compiler is complaining that you asked it to implicitly do something that the spec says must be an explicit cast :)

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+		return nil, fmt.Errorf("unable to push IP module: %v", err)
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
```
No need to repeat this comment each time, once on first use of the pattern is enough. Same comment throughout this file.

danderson commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+		return nil, fmt.Errorf("failed to set assignedName on Lifreq :%v", err)
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
```
For future readers, instead of these comments on integer casts, it'd be nice to have a comment for each ioctl block explaining what it does, for folks (like me) less familiar with the streams API.

danderson reviewed (commented):
Looking good, I'm glad all the unsafe syscall stuff is out of this codebase. Sprinkled a bit more feedback on style and re: the remaining XXXX questions. Thanks for continuing to work on this!

nshalman commented:
> Can you test this all and remove the XXX comments before I do a full review?

Yes. Thank you for your patience with me.

> If you have questions or need help with certain parts, feel free to say so in ordinary comments.

I vaguely understand what `routineRouteListener` is doing, but I don't fully understand how; it's the only function that has XXX comments remaining in it.

What is supposed to consume the events on the events channel and how do I confirm that they are being generated correctly? Does it have something to do with the ipc socket? How do I test the ipc socket? (Deleting it does make wireguard-go terminate, so that part seems to work, but I'm assuming it has some additional purpose?)

Also, a question about style. `go fmt` seems to have opinions about things like build tags being at the top of the file and which lines to put in there, while Dave is saying to put the license and copyright first and to remove one of them. Which one should I listen to?

zx2c4 commented:
> I vaguely understand what routineRouteListener is doing, but I don't fully understand how; it's the only function that has XXX comments remaining in it.

It's supposed to be monitoring for the interface going up, down, being removed (perhaps), or having its MTU changed.

> What is supposed to consume the events on the events channel and how do I confirm that they are being generated correctly?

Take a look at the readers of this channel for freebsd or openbsd, for example.

For testing, generate the events that it's supposed to be monitoring -- ifconfig .. up/down/mtu/destroy -- and see if this reflects in the program.

> Also, a question about style. go fmt seems to have opinions about things like build tags being at the top of the file and which lines to put in there, while Dave is saying to put the license and copyright first and to remove one of them. Which one should I listen to?

Do whatever the other files are doing. `go fmt` doesn't reformat the other files.

nshalman commented:
> > Also, a question about style. go fmt seems to have opinions about things like build tags being at the top of the file and which lines to put in there, while Dave is saying to put the license and copyright first and to remove one of them. Which one should I listen to?
> 
> Do whatever the other files are doing. `go fmt` doesn't reformat the other files.

Go 1.17 seems to have some new formatting opinions that Go 1.16 did not. I'll use 1.16 for formatting. Thanks!

zx2c4 commented:
Oh, good point. Fixed: https://git.zx2c4.com/wireguard-go/commit/?id=2ef39d47540c4537f0ddd3355fb95b33b91c09b7

nshalman commented on ipc/uapi_illumos.go:
```diff
@@ -0,0 +1,108 @@
+
+	go func(l *UAPIListener) {
+		for {
+			// start with lstat to avoid race condition
```
That comment was copied verbatim from the bsd (and linux) implementations.
I merely swapped in event ports to replace kqueue/kevent.

nshalman commented on main.go:
```diff
@@ -30,6 +30,7 @@ const (
 	ENV_WG_TUN_FD             = "WG_TUN_FD"
 	ENV_WG_UAPI_FD            = "WG_UAPI_FD"
 	ENV_WG_PROCESS_FOREGROUND = "WG_PROCESS_FOREGROUND"
+	ENV_WG_TUN_NAME           = "WG_TUN_NAME"
```
This is part of my cheating to avoid having to do complex stuff to identify which tun PPA we are using by examining the file descriptor. I've added some additional comments in the tun_illumos file to explain.
I don't know if @zx2c4 will let me get away with this cheating or not... :)

nshalman commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,355 @@
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
```
> Easiest way to check this would be to add some temporary logspam in the read loop

Any chance you could elaborate on how to do this?

nshalman reviewed (commented):
closing some threads

nshalman commented on tun/tun_illumos.go:
```diff
@@ -0,0 +1,381 @@
+	if err := unix.Putmsg(tun.fd, nil, buf[offset:], 0); err != nil {
+		return 0, err
+	}
+	return len(buf), nil
```
Should this be `len(buf[offset:])`?
(If `offset` is always `0`, it is fine, but it could be otherwise, right?)

zx2c4 replied:
Nice catch. It should indeed. 

You can make the first line of the function `buf = buf[offset:]`, and then just deal with buf itself the rest of the way through.

nshalman replied:
Fixed.

nshalman commented:
I haven't forgotten about this. @gco reached out to me to let me know that with a bit of copying functions around and s/illumos/solaris it was possible to get this code working on Solaris (My corresponding changes are here: https://github.com/nshalman/wireguard-go/compare/illumos-to-upstream...nshalman:wireguard-go:solaris). So https://go-review.googlesource.com/c/sys/+/437357 is open. When that is done, I hope to do some more cleanup of my [wg-quick bits](https://github.com/WireGuard/wireguard-tools/compare/master...nshalman:wireguard-tools:wg-quick-for-sunos) and get that all tested and squared away, but this has seen a lot of use by me and others, so hopefully I'll have it ready to land soon(ish).

zx2c4 commented:
Yalright, well, just let me know when it's finally done with no TODOs left, and I'm happy to review it.

nshalman commented:
> Yalright, well, just let me know when it's finally done with no TODOs left, and I'm happy to review it.

Looking forward to your review when you have the time. Thank you!!

nshalman commented:
Noting for myself...
This PR conflicts with #64. I tagged a version that I used with Tailscale 1.36 here: https://github.com/nshalman/wireguard-go/releases/tag/tailscale-1.36.0 / https://github.com/nshalman/wireguard-go/commit/0d59a42bb75d113d1ab95fbbbf7dcf66affb4c01

Not sure if @zx2c4 has a preference on which one lands first...
It also looks like this needs a merge conflict cleanup from me on `go.mod` and `go.sum`

zx2c4 commented:
I wasn't aware this was a serious effort, such that I should have sequencing concerns like that. It's been over two years now, so this is mostly your playground. But if you do think that the commit is done and mergable, please do pipe up so I can review and merge.

nshalman commented:
> I wasn't aware this was a serious effort, such that I should have sequencing concerns like that. It's been over two years now, so this is mostly your playground.

For someone smarter than me and with more free time I sure this could all have been done faster.

> But if you do think that the commit is done and mergable, please do pipe up so I can review and merge.

From a functionality perspective, all the testing that I've had the time and energy to do has shown this code to be complete.
I do have concerns about the related https://github.com/WireGuard/wireguard-tools/pull/17 having some subtle bug that I haven't found yet, but it too has stood up to my testing.

I did request a review, but I probably wasn't explicit enough, so that's on me (https://github.com/WireGuard/wireguard-go/pull/39#issuecomment-1293646049):

> Looking forward to your review when you have the time. Thank you!!

I will clean up the merge conflict and ping you very explicitly once that's done. I'll also mark the wireguard-tools PR ready for review and ping you on that one as well.

So. The next action is still on me and I will be very explicit in requesting review once I'm done with those steps.

Thank you for your attention!

nshalman commented:
@zx2c4 please review.

nshalman commented:
@jwhited your feedback would also be welcome.

nshalman commented:
Rebased to clear merge conflicts. Still ready for review by @zx2c4.

jwhited commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	return tun.events
+}
+
+func (tun *NativeTun) Read(bufs [][]byte, sizes []int, offset int) (int, error) {
```
`buffs` is used consistently elsewhere

nshalman replied:
Fixed. Thank you for the close read!

zx2c4 replied:
Darn. I was rushing home to beat you to it but...

https://git.zx2c4.com/wireguard-go/commit/?id=0ad14a89f5f9da577dae6a63ad196015e51a0381

`bufs` seems more correct to me, so I fixed it everywhere.

nshalman replied:
Ha! I will "unfix" it.

zx2c4 commented on ipc/uapi_solaris.go:
```diff
@@ -0,0 +1,111 @@
+				l.connErr <- err
+				return
+			}
+			_, perr := l.evPort.GetOne(nil)
```
`_, err = l.evPort.GetOne(nil)`

zx2c4 commented on ipc/uapi_solaris.go:
```diff
@@ -0,0 +1,111 @@
+	}(uapi)
+
+	// watch for new connections
+
```
No need for this new line.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	defer unix.Close(routeSocket)
+
+	data := make([]byte, os.Getpagesize())
+	for {
```
Can you confirm that this loop works in a variety of circumstances?

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+		// a chance to notice
+		// 2. It doesn't really respect 'ifconfig tunN down' properly so this code
+		// won't notice, and it will be broken and refuse to come back up
+		// if it is set down.
```
Not sure I understand what you mean. It'll get set down but won't notice coming back up? It won't notice going down but it'll come back up just fine?

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+		}
+		defer unix.Close(fd)
+
+		// get the STREAMS muxid for the tun device stream
```
"get" -- what are you getting? You're putting a constant into a variable, casting it to an int. And then later...

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+
+		// solaris/illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint.
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
```
...and then later you make it into a uint. Why?

Just write `IoctlLifreq(fd, uint(unix.SIOG...), &l)`

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	assignedName := fmt.Sprintf("tun%d", ppa)
+
+	// Clean up anything possibly left behind by a crash
+	tunDestroy(assignedName)
```
Interesting...

What have you experienced being left behind? What are the circumstances here?

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	tunDestroy(assignedName)
+
+	// The solaris/illumos tun driver doesn't have an ioctl for reading back the PPA.
+	// Just pass it through to the daemon process via environment variable.
```
Darn. Still? Weren't you talking to them about adding this at some point?

zx2c4 replied:
Where's the source code for the tun driver? Is it not part of the base illumos?

nshalman replied:
> Weren't you talking to them about adding this at some point?

No, that was other functionality. 

Also, I'm told that the code as it is today works on Solaris as well, an operating system that is closed source and over which I have even less influence. As gross as this is, I think it's a decent tradeoff to make to be able to support Solaris too.

If anything, some day illumos might get a newer better TUN driver that doesn't even use STREAMS and the solaris and illumos implementations in here would diverge. But that might not happen until I learn enough C to write one (and find the time and motivation to do so), so that could be (even more) years away...

zx2c4 replied:
Can you link to the source code of the tun driver in illumos?

nshalman replied:
As linked in a much older thread, the header file is here:
https://github.com/kaizawa/tuntap/blob/v1.3.3/if_tun.h

(Linking the version string referenced in e.g. the OmniOS build recipe:
https://github.com/omniosorg/omnios-build/blob/master/build/tuntap/build.sh#L21)

@gco or @alanc might be able to confirm that Solaris is using the same version...

gco replied:
Solaris does not include the tuntap driver AFAIK, end-users need to install it if they need it.    @danmcd would probably know the history (and merged tuntap into illumos!).

nshalman replied:
> Solaris does not include the tuntap driver AFAIK, end-users need to install it if they need it. @danmcd would probably know the history (and merged tuntap into illumos!).

It's not in-gate in illumos. The distros ship it separately.

I linked to the OmniOS build recipe above. The SmartOS one is https://github.com/TritonDataCenter/illumos-extra/tree/master/tun and also references 1.3.
The OpenIndiana one is at https://github.com/OpenIndiana/oi-userland/tree/oi/hipster/components/network/tuntap

I don't see the Solaris build recipe in https://github.com/oracle/solaris-userland/tree/master/components so I'm guessing that it's in a closed repo. I was just wondering what version string you folks have and whether you've made any interesting changes to it.

gco replied:
I'll emphasize "does not include".   Solaris users need to get it from the 10 year old github repository.

nshalman replied:
> I'll emphasize "does not include". Solaris users need to get it from the 10 year old github repository.

Got it. So to be very explicit, in order to use my wireguard-go and tailscale ports you installed the driver directly from the https://github.com/kaizawa/tuntap repo. Is that correct?

gco replied:
I got mine from @danmcd many years ago as part of a private tool called punchin.  It's built from the same sources.

zx2c4 replied:
Okay so it sounds like there's a single (old & crusty) codebase for tuntap shared across Solaris, Illumos, and derivatives. So,

> Also, I'm told that the code as it is today works on Solaris as well, an operating system that is closed source and over which I have even less influence. As gross as this is, I think it's a decent tradeoff to make to be able to support Solaris too.

This is not correct then, right?

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+
+	// The "tun" device is a STREAMS module.  We need to make a STREAMS ioctl
+	// request to that module using the TUNNEWPPA command.  Returns the newly
+	// allocated PPA number. Passing -1 will request a dynamically assinged
```
Where's the source code for the tun driver? Is it not part of the base illumos?

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	var ppa int
+	_, err := fmt.Sscanf(tun.name, "tun%d", &ppa)
+	if err != nil {
+		return fmt.Errorf("received a bogus tun name: %s", tun.name)
```
s/bogus/invalid/
s/received a //

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+		return fmt.Errorf("received a bogus tun name: %s", tun.name)
+	}
+
+	// Unclear why, but we need to open a second handle on the tun device for
```
Figure out why and document it.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+
+	// Unclear why, but we need to open a second handle on the tun device for
+	// the ioctls that follow. Reusing the existing one does not work.
+	tunFD, err := unix.Open(tunNode, unix.O_RDWR, 0)
```
Everywhere else in the codebase, the D in Fd is lowercase.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+
+	// select which ppa we're using
+	req := int(unix.IF_UNITSEL)
+	err = unix.IoctlSetPointerInt(tunFD, uint(req), ppa)
```
No need to make req a separate variable.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	// By holding open this file descriptor on /dev/ip for the life of the connection
+	// and using I_LINK, STREAMS will tear everything down cleanly when we close these
```
But then I_LINK isn't actually set until deep down in a separate function. Maybe configureSTREAMS should be inlined into this function instead of being standalone, so that it's more clear.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+
+	var l unix.Lifreq
+	if err := l.SetName(tun.name); err != nil {
+		return fmt.Errorf("failed to set name on Lifreq :%v", err)
```
`failed to` here, but `unable to` elsewhere. Look at what wireguard-go does elsewhere and copy that.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	// set the IP muxid
+	reqnum := int(unix.SIOCSLIFMUXID)
```
reqnum doesn't need to be a variable.

zx2c4 commented on tun/tun_solaris.go:
```diff
@@ -0,0 +1,398 @@
+		return err
+	}
+
+	reqnum := int(unix.SIOCSLIFMTU)
```
reqnum doesn't need to be a separate variable.

gco replied:
See also: https://go.googlesource.com/sys/+/master/unix/syscall_solaris_test.go#370

nshalman replied:
> See also: https://go.googlesource.com/sys/+/master/unix/syscall_solaris_test.go#370

Dunno if that counts; I wrote it. lolsob

zx2c4 replied:
The compiler isn't confused; you are. Your comment in that test isn't accurate and does a disservice.

It looks like ioctl has the wrong signature in Go. According to https://illumos.org/man/2/ioctl req is an int. In Go, that's int32, not int or uint as the Go code has now. Presumably it's too late to fix that, which is unfortunate.

What you can do, though, is just avoid the negative ioctl constant for the ones you added, which probably have no users yet anyway. This way you don't run into ambiguity with sign extension. The easiest way to do this is probably just by adding `& 0xffffffff` onto the end of the constant definition. While you're at it, you can fix your bogus comment in the test too.

Be sure to submit this to x/sys with a compelling argument that you understand well; don't just link to this comment or something and call it a day.

zx2c4 commented:
From your top message:

> reliability testing via Tailscale (Tailscale for both Solaris and illumos can be built on top of this work and it has been quite reliable.)

Testing with third-party software is not sufficient. Test `wireguard-go` itself and then report back. Please write what you've tested and how.

zx2c4 commented:
> Do you have a list of specific tests you would want to see done?

I'm primarily concerned about all the weird cases with event listers on the unix socket, tun interface, and all that. So it'd be nice to know that you've printf'd up those code paths at some point to make sure its the behavior you're expecting. The lengthy development here and unsatisfactory earlier results make me slightly more wary of code that looks fine but might not be. So please just make sure that all parts of this PR are intentional and understood, and test that they're doing what that understanding expects them to do.

zx2c4 commented:
I just spent a while reading through all the review over the last 3 years and the various revisions and such. I think I'm uncomfortable with this landing through Github and don't have enough trust that you've dotted all your is and crossed your ts. So I think we need to move to a more rigid and less haphazard review process.

Please submit this to the wireguard mailing list when it's ready, as "PATCH v1", with a proper, detailed, and well-written commit message that explains the whole patch, the testing you've done, the provenance of the tun driver this is written for, rudimentary performance measurements, and so on. Be sure to add to the `Cc: ...` trailers relevant people, such as those various Illumos developers and maybe the author of the tun driver, and whoever else you feel should be there. I'll want to see a review in that thread from somebody who knows Illumos, so if you know anybody there, be sure to have them in the Cc also.

Be sure to read the documentation on submitting patches to mailing lists in the Linux kernel repository.

Yes, this is annoying and probably feels like paperwork. But there's a lot of inconsistency here, and I don't feel alright moving forward with the usual "it's probably fine" stance I ordinarily have.
